video/out/opengl/context.c

```diff
@@ -103,140 +34,286 @@ const int mpgl_preferred_gl_versions[] = {
+    struct ra_gl_ctx *p = ctx->gl_priv;
+    GL *gl = p->gl;
+
+    if (p->opts->use_glfinish)
+        gl->Finish();
+
+    if (gl->FenceSync && !gl->external_swchain) {
+        GLsync fence = gl->FenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0);
+        if (fence)
+            MP_TARRAY_APPEND(ctx, p->vsync_fences, p->num_vsync_fences, fence);
+    }
+
+    switch (p->opts->early_flush) {
+    case FLUSH_AUTO:
+        if (frame->display_synced)
+            break;
+        // fall through
+    case FLUSH_YES:
+        gl->Flush();
+    }
+
+    return true;
 }
 
-void mpgl_start_frame(struct MPGLContext *ctx)
+static void check_pattern(struct ra_ctx *ctx, int item)
 {
-    if (ctx->driver->start_frame)
-        ctx->driver->start_frame(ctx);
+    struct ra_gl_ctx *p = ctx->gl_priv;
+    int expected = p->opts->vsync_pattern[p->last_pattern];
+    if (item == expected) {
+        p->last_pattern++;
+        if (p->last_pattern >= 2)
+            p->last_pattern = 0;
+        p->matches++;
+    } else {
+        p->mismatches++;
+        MP_WARN(ctx, "wrong pattern, expected %d got %d (hit: %d, mis: %d)\n",
+                expected, item, p->matches, p->mismatches);
+    }
 }
 
-void mpgl_swap_buffers(struct MPGLContext *ctx)
+void ra_gl_ctx_post_swap_buffers(struct ra_ctx *ctx)
 {
-    ctx->driver->swap_buffers(ctx);
+    struct ra_gl_ctx *p = ctx->gl_priv;
+    GL *gl = p->gl;
+
+    p->frames_rendered++;
+    if (p->frames_rendered > 5 && !ctx->opts.debug)
+        ra_gl_set_debug(ctx->ra, false);
+
+    if ((p->opts->waitvsync || p->opts->vsync_pattern[0])
+        && gl->GetVideoSync)
+    {
+        unsigned int n1 = 0, n2 = 0;
+        gl->GetVideoSync(&n1);
+        if (p->opts->waitvsync)
+            gl->WaitVideoSync(2, (n1 + 1) % 2, &n2);
+        int step = n1 - p->prev_sgi_sync_count;
+        p->prev_sgi_sync_count = n1;
+        MP_DBG(ctx, "Flip counts: %u->%u, step=%d\n", n1, n2, step);
+        if (p->opts->vsync_pattern[0])
+            check_pattern(ctx, step);
+    }
+
+    while (p->num_vsync_fences >= ctx->opts.swchain_depth) {
+        gl->ClientWaitSync(p->vsync_fences[0], GL_SYNC_FLUSH_COMMANDS_BIT, 1e9);
+        gl->DeleteSync(p->vsync_fences[0]);
+        MP_TARRAY_REMOVE_AT(p->vsync_fences, p->num_vsync_fences, 0);
+    }
 }
 
-void mpgl_uninit(MPGLContext *ctx)
+struct mp_image *ra_gl_ctx_screenshot(struct ra_ctx *ctx)
 {
-    if (ctx)
-        ctx->driver->uninit(ctx);
-    talloc_free(ctx);
+    struct ra_gl_ctx *p = ctx->gl_priv;
+    assert(p->wrapped_fb);
+
+    struct mp_image *screen = gl_read_fbo_contents(p->gl, p->main_fb,
+                                                   p->wrapped_fb->params.w,
+                                                   p->wrapped_fb->params.h);
+
+    // OpenGL FB is also read in flipped order, so we need to flip when
+    // the rendering is *not* flipped. I hope that made sense
+    if (!ctx->flip_v)
+        mp_image_vflip(screen);
```

Comment:
screen can be NULL on failure, and the function doesn't allow NULL.

Reply:
fixed